src/posix/waiter.cpp

```diff
@@ -26,11 +26,20 @@ dstime Waiter::ds;
 
 PosixWaiter::PosixWaiter()
 {
-    // pipe to be able to leave the select() call
-    if (pipe(m_pipe) < 0)
+    for (;;)
     {
-        LOG_fatal << "Error creating pipe";
-        throw std::runtime_error("Error creating pipe");
+        // pipe to be able to leave the select() call
+        if (pipe(m_pipe) < 0)
+        {
+            LOG_fatal << "Error creating pipe";
+            throw std::runtime_error("Error creating pipe");
+        }
+
+        // On MacOS, pipe() keeps giving us 3 as the first fd.  Possibly on the first pipe() call per thread?
+        if (m_pipe[0] > 3 && m_pipe[1] > 3) break;
+
+        if (m_pipe[0] > 3) close(m_pipe[0]);
+        if (m_pipe[1] > 3) close(m_pipe[1]);
```

Comment:
:thinking: I feel we could leak the `fd=3` and keep it opened forever. If first call to pipe returns 3 and 4, only the 4 will be closed. Is it the purpose? so you don't get the `3` anymore in subsequent calls?
Looks like a nasty bug to return the same `fd` in result of different calls (without closing it before).

Reply:
Yes the intention was to leak the fd 3 so that we don't accidentally reuse it - however we have a much better explanation and fix now, so we can abandon this PR (for now at least, unless some other issue arises with the waiter on Mac)  thanks